source/hands_on/d4pg.rst

```diff
@@ -0,0 +1,110 @@
+D4PG
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Overview
+---------
+D4PG, proposed in the paper `Distributed Distributional Deterministic Policy Gradients <https://arxiv.org/abs/1804.08617v1>`_,
+is an actor-critic, model-free policy gradient algorithm that extends upon `DDPG <https://arxiv.org/abs/1509.02971>`_.
+Improvements over DDPG include the use of N-step returns, prioritized experience replay and distributional value function.
+Moreover, training is parallelized with multiple distributed workers all writing into the same replay table.
+The authors found that these simple modifications contribute to the overall performance of the algorithm with N-step
+returns brining the biggest performance gain and priority buffer being the less crucial one.
+
+Quick Facts
+-----------
+1. D4PG is only used for environments with **continuous action spaces**.(i.e. MuJoCo)
+
+2. D4PG is an **off-policy** algorithm.
+
+3. D4PG uses a **distributional** critic.
+
+4. D4PG is a **model-free** and **actor-critic** RL algorithm, which optimizes actor network and critic network, respectively.
+
+5. Usually, D4PG use **Ornstein-Uhlenbeck process** or **Gaussian process** (default in our implementation) for exploration.
+
+Key Equations or Key Graphs
+---------------------------
+The D4PG algorithm maintains a distributional critic :math:`Z_\pi(x, a)` and the loss used to learn the distribution parameters is defined as :math:`L(\pi) = \mathbb{E}[d(\mathcal{T}_{\mu_\theta}, Z_{pi'}(s, a), Z_\pi(s, a)]`, where :math:`\mathcal{T}_{\mu_\theta}` si the Bellman operator. The implemented distribution is usually a Categorical distribution.
+
+The **action-value distribution** inside the actor update is done by taking the expectation with respect to the action-value distribution:
+
+.. math::
+    \begin{aligned}
+    \nabla_\theta J(\theta)
+    &\approx \mathbb{E}_{\rho^\mu} [\nabla_a Q_w(s, a) \nabla_\theta \mu_\theta(s) \rvert_{a=\mu_\theta(s)}] \\
+    &= \mathbb{E}_{\rho^\mu} [\mathbb{E}[\nabla_a Z_w(s, a)] \nabla_\theta \mu_\theta(s) \rvert_{a=\mu_\theta(s)}]
+    \end{aligned}
+
+When calculating the TD error, **N-step** incorporate in the TD target rewards in more future steps:
+
+.. math::
+    r(s_0, a_0) + \mathbb{E}[\sum_{n=1}^{N-1} r(s_n, a_n) + \gamma^N Q(s_N, \mu_\theta(s_N)) \vert s_0, a_0 ]
+
+D4PG samples from a **prioritized replay buffer** with a non-uniform probability :math:`p_i`.
+This requires the use of importance sampling, implemented by weighting the critic update by a factor of :math:`R_{p_i}^{-1}`.
+
+
+Pseudocode
+----------
+
+.. image::
+   images/D4PG_algo.png
+   :scale: 75%
+   :align: center
+
+
+Implementations
+----------------
+The default config is defined as follows:
+
+.. autoclass:: ding.policy.d4pg.D4PGPolicy
+   :noindex:
```

Comment:
add more comments in d4pg policy config?

Reply:
I think the 'config' section is not shown because of a format problem in di-engine. I will investigate it and change accordingly.